pkg/controller/elasticsearch/pdb/roles.go

```diff
@@ -0,0 +1,518 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License 2.0;
+// you may not use this file except in compliance with the Elastic License 2.0.
+
+package pdb
+
+import (
+	"context"
+	"fmt"
+	"slices"
+	"sort"
+
+	appsv1 "k8s.io/api/apps/v1"
+	policyv1 "k8s.io/api/policy/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/client-go/kubernetes/scheme"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	esv1 "github.com/elastic/cloud-on-k8s/v3/pkg/apis/elasticsearch/v1"
+	"github.com/elastic/cloud-on-k8s/v3/pkg/controller/common/metadata"
+	"github.com/elastic/cloud-on-k8s/v3/pkg/controller/elasticsearch/label"
+	"github.com/elastic/cloud-on-k8s/v3/pkg/controller/elasticsearch/sset"
+	"github.com/elastic/cloud-on-k8s/v3/pkg/utils/k8s"
+	"github.com/elastic/cloud-on-k8s/v3/pkg/utils/set"
+)
+
+var (
+	// group the statefulsets by the priority of their roles.
+	// master, data_*, ingest, ml, transform, coordinating, and we ignore remote_cluster_client as it has no impact on availability
+	priority = []string{"master", "data", "data_frozen", "ingest", "ml", "transform", "coordinating"}
```

Comment:
Should we try to reuse the types and logic from [https://github.com/elastic/cloud-on-k8s/blob/main/pkg/apis/elasticsearch/v1/elasticsearch_config.go](pkg/apis/elasticsearch/v1/elasticsearch_config.go)? It feels like there is a bit a of duplicated logic in some places, for example in `normalizeRole`  which is almost already implemented in [CanContainData()](https://github.com/elastic/cloud-on-k8s/blob/1832f4bbc4e280ffeff5df96fd4ce3bf503899b4/pkg/apis/elasticsearch/v1/elasticsearch_config.go#L62)

```suggestion
	priority = []esv1.NodeRole{esv1.MasterRole, esv1.DataRole, esv1.DataFrozenRole, esv1.IngestRole, esv1.MLRole, esv1.TransformRole, esv1.CoordinatingRole}
```